### DISCUSSION TITLE
[MongoDB] What would be the best way to create reports about user actions?

### DISCUSSION THREAD

**Discussion by bn-piar** [prisma/prisma#16626] [Q&A]
Hi guys. I'm working on a warehouse app where user can add/edit/delete products to/from warehouse and I need to "log" all user actions, for example what user deleted a product, what was the product details (such as name, barcode, etc.)

I'm currently creating reports in middleware `prisma.$use` but many actions and data details overlap, it's very hard to maintain/add new features and we are experiencing a lot of nasty bugs related to middleware because of actions and data overlaps. 

Also in dev mode (next.js) for some reason middleware is triggered multiple times: when product is created/changed/deleted, multiple same reports are created. 

I was thinking to move report related logic to corresponding product handlers, so `createProductHandler` would also create a report for product creation, `updateProductHandler` would create a report for product update, etc., but I'm not sure if it's the way to go or there are better ways.

Is there a best practice for this kind of thing? How would you do that?

**Comment by nurul3101** [comment]
Hello @bn-piar 👋 

I would recommend you to have a look at [Prisma Client Extensions](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions) which were released recently in Prisma 4.7.0 release.
You can extend Prisma Client and add your [custom methods](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/model#add-a-custom-method) which perform these Audit Log operations, and you can invoke these methods from the product handlers or from any other location. This would make it much easier to manage these operations and would be much cleaner compared to using Middlewares.

**Reply by bn-piar** [reply]
I can't use prisma client extension in my app because of [Maximum call stack size exceeded](https://github.com/prisma/prisma/issues/16600) error

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
